src/string_id.h

```diff
@@ -69,69 +107,92 @@ template<typename T>
 class string_id
 {
     public:
+        template<class U>
+        friend class string_id;
         using This = string_id<T>;
 
-        /**
-         * Forwarding constructor, forwards any parameter to the std::string
-         * constructor to create the id string. This allows plain C-strings,
-         * and std::strings to be used.
-         */
-        // Beautiful C++11: enable_if makes sure that S is always something that can be used to constructor
-        // a std::string, otherwise a "no matching function to call..." error is generated.
+        template<unsigned N>
+        explicit constexpr string_id( const char ( &str )[N] ) : _hash( Hash::CompileTime::fnv1_64( str ) ),
+            _char_ptr( str ), _is_char_ptr( true ) {}
+
+        template<typename U>
+        string_id( const string_id<U> &other ) : _hash( other.hash() ), _char_ptr( other._char_ptr ),
+            _is_char_ptr( other._is_char_ptr ) {}
+
         template<typename S, class = typename
-                 std::enable_if< std::is_convertible<S, std::string >::value>::type >
-        explicit string_id( S &&
-                            id ) : _id( std::forward<S>( id ) ), _cid( INVALID_CID ), _version( INVALID_VERSION ) {}
+                 std::enable_if< std::is_convertible< S, std::string >::value >::type
+                 >
+        explicit string_id( S && id ) {
+            auto str = std::make_unique<std::string>( std::forward<S>( id ) );
+            uint64_t hash = Hash::RunTime::fnv1_64( str->c_str(), str->size() );
+            _hash = hash;
+
+            auto insertion = cached_strings().try_emplace( hash, std::move( str ) );
```

Comment:
Does `cached_strings` ever gets cleared? What will happen if somebody repeatedly generates short lived dynamic string ids?

Reply:
It does not get cleared, but why should it? The only case where you'd want that is if there's an ID that is generated once and never used again. As most IDs do not act this way, I think the performance cost of making this an lru wouldn't be worth it
As for short lived ids - that shouldn't be a problem. All of the strings are pooled, so there will only be a single copy in memory which will be referenced by all ids.

Reply:
Even if currently there is no situation where string_ids are used in dynamic setting (i.e. unique ids are repeatedly generated and then discarded), your implementation means that the _semantics_ of string_ids is changing. I.e. currently it doesn't leak memory, with your approach it might. In my opinion this is an important issue that should be carefully considered.